### PR TITLE
feature: add support for mpijob in katib

### PR DESCRIPTION
**Pull request by YuxiJin-tobeyjin**
**What this PR does / why we need it**:
Add support for mpijob in katib

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes  https://github.com/kubeflow/katib/issues/1181

**Release note**:

```release-note
None
```

cc @gaocegege @johnugeorge @andreyvelich

### CODE REVIEW

**Comment by YuxiJin-tobeyjin** [comment]
@googlebot I signed it!

File: pkg/controller.v1alpha3/consts/const.go [hunk 1]
```diff
@@ -112,11 +112,14 @@ const (
...
 	// JobKindPyTorch is the kind of PyTorchJob.
 	JobKindPyTorch = "PyTorchJob"
+	// JobKindMpi is the kind of MpiJob.
+	JobKindMpi = "MPIJob"
 
 	// built-in JobRoles
 	JobRole        = "job-role"
 	JobRoleTF      = "tf-job-role"
 	JobRolePyTorch = "pytorch-job-role"
+	JobRoleMpi     = "mpi_role_type"
```
[the remark this answers is not in the document]
Comment (YuxiJin-tobeyjin) [reply pkg/controller.v1alpha3/consts/const.go:122]: Yes, refer to 
<https://github.com/kubeflow/mpi-operator/blob/master/pkg/controllers/v1alpha2/mpi_job_controller.go#L81>

File: pkg/metricscollector/v1alpha3/common/const.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ const (
 
 	TrainingCompleted = "completed"
 
-	DefaultFilter = `([\w|-]+)\s*=\s*((-?\d+)(\.\d+)?)`
+	DefaultFilter = `([\w|-]+)\s*[:=]\s*((-?\d+)(\.\d+)?)`
```
[the remark this answers is not in the document]
Comment (YuxiJin-tobeyjin) [reply pkg/metricscollector/v1alpha3/common/const.go:33]: In most cases ,the print logs are like “filtername : *%“ or “filtername = *%“, so  I prefer the default filter to support both “:” and "=",  would it be better? @andreyvelich WDYT

File: pkg/controller.v1alpha3/consts/const.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/controller.v1alpha3/consts/const.go:122]: @terrytangyuan Hi, will we change the label in the future, maybe v1beta1 or v1?

File: pkg/metricscollector/v1alpha3/common/const.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sperlingxx) [reply pkg/metricscollector/v1alpha3/common/const.go:33]: I'm happy with this modification.

File: pkg/controller.v1alpha3/consts/const.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreyvelich) [reply pkg/controller.v1alpha3/consts/const.go:122]: Agree with @gaocegege. I think MPI should follow the same patterns as TF Job and Pytorch Job.

File: pkg/metricscollector/v1alpha3/common/const.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreyvelich) [reply pkg/metricscollector/v1alpha3/common/const.go:33]: If we change this, I think we should modify documentation about default filter for Metrics Collector. For example here: https://www.kubeflow.org/docs/components/hyperparameter-tuning/experiment/#metrics-collector.

Do you think specifying filter with filter format (:) will be not handy for Katib user?
E.g, like we did for DARTS example: https://github.com/kubeflow/katib/blob/master/examples/v1alpha3/nas/darts-example-gpu.yaml#L19.
Your thoughts @gaocegege @johnugeorge ?

File: pkg/controller.v1alpha3/consts/const.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (terrytangyuan) [reply pkg/controller.v1alpha3/consts/const.go:122]: Thanks. I changed it in https://github.com/kubeflow/mpi-operator/pull/252 for v1 candidate of MPI Operator. Perhaps this PR can add support for v1 candidate directly? The API should be relatively stable even though there isn't official release yet.

File: pkg/webhook/v1alpha3/pod/inject_webhook_test.go [hunk 3]
```diff
@@ -480,6 +480,23 @@ func TestGetKatibJob(t *testing.T) {
 			Err:  true,
 			Name: "Invalid Kind",
 		},
+		{
+			Pod: v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: "batch/v1",
+							Kind:       "Job",
```
Comment (terrytangyuan) [inline pkg/webhook/v1alpha3/pod/inject_webhook_test.go:489]: FYI we removed Job and Statefulset for v1 candidate here so now MPI Operator only creates pods: https://github.com/kubeflow/mpi-operator/pull/203

File: pkg/controller.v1alpha3/consts/const.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreyvelich) [reply pkg/controller.v1alpha3/consts/const.go:122]: Thank you @terrytangyuan. I agree, let's support directly v1 MPI Operator in Katib.

[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/controller.v1alpha3/consts/const.go:122]: SGTM

File: pkg/metricscollector/v1alpha3/common/const.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/metricscollector/v1alpha3/common/const.go:33]: I think we can keep it to enhance the default filter. But we need to open an issue to update the doc.

File: pkg/controller.v1alpha3/consts/const.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (YuxiJin-tobeyjin) [reply pkg/controller.v1alpha3/consts/const.go:122]: OK, actually I have my concern, release branch is always much more stable than dev branch, thus I choose to support the latest release, perhaps I should wait for the next official release for mpi-operator and then submit a new PR? @gaocegege @andreyvelich WDYT

File: pkg/webhook/v1alpha3/pod/inject_webhook_test.go (hunk 3 above)
Comment (YuxiJin-tobeyjin) [reply pkg/webhook/v1alpha3/pod/inject_webhook_test.go:489]: (⊙o⊙)…  we developed some features based on worker's sts property...

File: pkg/controller.v1alpha3/consts/const.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/controller.v1alpha3/consts/const.go:122]: I think katib will release new version after MPI operator releases its next version. Thus I think we can safely support v1 to avoid additional work. /cc @johnugeorge

File: pkg/webhook/v1alpha3/pod/inject_webhook_test.go (hunk 3 above)
Comment (YuxiJin-tobeyjin) [reply pkg/webhook/v1alpha3/pod/inject_webhook_test.go:489]: @terrytangyuan Is there any plan when to release a new mpi-operator version?

File: pkg/controller.v1alpha3/consts/const.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (terrytangyuan) [reply pkg/controller.v1alpha3/consts/const.go:122]: The MPI v1 controller and API should be relatively stable now so it's better to just support that.

File: pkg/webhook/v1alpha3/pod/inject_webhook_test.go (hunk 3 above)
Comment (terrytangyuan) [reply pkg/webhook/v1alpha3/pod/inject_webhook_test.go:489]: No but it should be relatively stable now as we are getting close to graduation.

File: pkg/controller.v1alpha3/consts/const.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andreyvelich) [reply pkg/controller.v1alpha3/consts/const.go:122]: Agree with @gaocegege . 
@YuxiJin-tobeyjin can you change this PR to support v1 MPI Operator version?

[the remark this answers is not in the document]
Comment (YuxiJin-tobeyjin) [reply pkg/controller.v1alpha3/consts/const.go:122]: Ok, but I need to make some tests on latest mpi master branch to ensure changes work as expected, I will do it ASAP.

[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/controller.v1alpha3/consts/const.go:122]: Thanks for your contribution! :tada: :+1:

[the remark this answers is not in the document]
Comment (YuxiJin-tobeyjin) [reply pkg/controller.v1alpha3/consts/const.go:122]: @terrytangyuan  I've tried mpi-operator master branch on our cluster with kubernetes 1.14 and encountered some problems.

First, when creating new mpijob crd ,it throws up

```log
unknown field "additionalPrinterColumns" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceValidation
```

Walk around then, the crd has been created ok, but the mpi-operator is created failed 

```log
the server could not find the requested resource (get mpijobs.kubeflow.org)
```
Maybe problem is here ,still use KubeflowV1alpha2 to get v1 crd?

```url
https://github.com/kubeflow/mpi-operator/blob/acddf3028ce922e24ac3e735a2928ff4487be28f/cmd/mpi-operator.v1/app/server.go#L288
```
Walk around then, but mpi-operator throws 

```log
Failed to list *v1beta1.PodGroup: podgroups.scheduling.volcano.sh
```
So , now use mpi-operator v1 we must use volcano as the bacth scheduler? Since kube-batch latest release do not support v1beta1.PodGroup.  Actually we've already used the kube-batch latest release for some times...

[the remark this answers is not in the document]
Comment (terrytangyuan) [reply pkg/controller.v1alpha3/consts/const.go:122]: Thanks! I fixed the CRD validation issue in https://github.com/kubeflow/mpi-operator/pull/257#issuecomment-628612127. Yes Volcano has better support and maintenance. You may want to update your MPI Operator's deployment yaml which includes permissions for Volcano resources: https://github.com/kubeflow/mpi-operator/blob/master/deploy/v1/mpi-operator.yaml

**Comment by carmark** [comment]
@andreyvelich The proposal was merged, could we move this PR forward or submit a new one?

**Comment by YuxiJin-tobeyjin** [comment]
Close cause https://github.com/kubeflow/katib/pull/1273 will give a better solution.

**Comment by andreyvelich** [comment]
@carmark Sure, I have already created PR to support MPI in Katib: https://github.com/kubeflow/katib/pull/1342.
